### ISSUE TITLE
Submit and merge the PlayStation 2 Linux kernel port

### ISSUE THREAD

**Issue by frno7** [frno7/linux#1]
An initial patch submission of a PlayStation 2 Linux port to the kernel mailing list would most likely include USB OHCI and Graphics Synthesizer device drivers, as they are needed for a reasonable level of system usability.

See also #2, #5 and #30.

**Comment by frno7** [comment]
The [initial PlayStation 2 Linux kernel patch submission](https://lore.kernel.org/linux-mips/cover.1567326213.git.noring@nocrew.org/T/) has now been posted to the MIPS Linux kernel mailing list.

**Comment by frno7** [comment]
A note that Paul Burton, who indicated he would [look through the patch submission in detail soon](https://lore.kernel.org/linux-mips/20190904141026.hwzibvauis5sizq6@pburton-laptop/), is [no longer working for MIPS the company](https://lkml.org/lkml/2020/2/22/206):

> My time with MIPS the company has reached its end, and so at best I'll have little time spend on maintaining arch/mips/.
> 
> Ralf last authored a patch over 2 years ago, the last time he committed one is even further back & activity was sporadic for a while before that. The reality is that he isn't active.
> 
> Having a new maintainer with time to do things properly will be beneficial all round. Thomas Bogendoerfer has been involved in MIPS development for a long time & has offered to step up as maintainer, so add Thomas and remove myself & Ralf from the MIPS entry.

**Comment by frno7** [comment]
32-bit architectures will [likely become obsoleted](https://lwn.net/Articles/1035727/) in future Linux kernels. Although the R5900 has 64-bit features, it’s primarily regarded as a 32-bit processor. Framebuffers and other key pieces of the Linux kernel are also being degraded, or removed, as it seems. Hence, the best home for PlayStation 2 Linux may well be the 5.x series.

**Comment by AKuHAK** [comment]
Just for the information: https://github.com/llvm/llvm-project/pull/176666

**Comment by frno7** [comment]
Thanks, @AKuHAK! It’s obviously very good to have more than one compiler available for the R5900. :-)

**Comment by rickgaiser** [comment]
In case you're intertested, I've got a full featured LLVM compiler here:
https://github.com/rickgaiser/llvm-project/tree/EmotionEngine
https://github.com/rickgaiser/llvm-project/blob/EmotionEngine/PS2_EE_README.md

All instructions (EE, MMI, VU0 macro, etc...) with full autovectorization implemented. I plan to merge upstream in very small pieces.

Also allows for more fun things on the ps2, like new programming languages for instance:
https://github.com/rickgaiser/test_llvm_zig

**Comment by frno7** [comment]
Impressive, @rickgaiser! Is `-mabi=o32` within reach too, as that’s what we currently have for Linux and QEMU? (n32 for Linux needs some work on the kernel, but should be possible, eventually.)

**Comment by rickgaiser** [comment]
mips on llvm is nothing new. I only use n32 and that was supported out of the box, I assume o32 is no different. Only change I did for n32 is add the VU0 registers, but that was optional. I assume the VU0 works for o32 just fine, but using the stack instead of registers.

**Comment by frno7** [comment]
@AKuHAK, @rickgaiser, I’ve opened #90 for assembler, compiler, tooling and file format (ELF, etc.) discussions!

**Comment by mirh** [comment]
> Framebuffers and other key pieces of the Linux kernel are also being degraded, or removed, as it seems.

Weren't you all zesty and hopeful in #30 for their resurrection? The amiga is still getting updates nowadays, and in x86 land they just barely got to remove i486.

**Comment by frno7** [comment]
> Weren't you all zesty and hopeful in [#30](https://github.com/frno7/linux/issues/30) for their resurrection?

Keep in mind that issue #30 was registered in 2019, which was seven years ago. The Linux [DRM](https://en.wikipedia.org/wiki/Direct_Rendering_Manager) subsystem turned out to require [EDID](https://en.wikipedia.org/wiki/Extended_Display_Identification_Data), which the PlayStation 2 doesn’t support, among other things. The DRM isn’t a good fit for the PlayStation 2, because it’s primarily intended for modern GPUs. See https://github.com/frno7/linux/issues/10#issuecomment-586567308 and https://github.com/frno7/linux/issues/22#issuecomment-1019426768.

> The amiga is still getting updates nowadays, and in x86 land they just barely got to remove i486.

The [Amiga](https://en.wikipedia.org/wiki/Amiga) is fundamentally a much simpler machine than the PlayStation 2, so one might expect it to hang around for a while yet. Even so, the Amiga is 32-bit, and therefore [likely to be removed](https://lwn.net/Articles/1035727/) in forthcoming Linux kernels, along with all other 32-bit architectures.

New Linux kernel features, such as [page folios](https://lwn.net/Articles/849538/) and [`io_uring`](https://en.wikipedia.org/wiki/Io_uring), are not particularly useful for the PlayStation 2, due to lack of memory, etc. With no compelling new features in recent kernels and increasing 32-bit subsystem breakage and deprecation, I think staying on Linux 5.4.x is better than attempting recent kernels.

**Comment by mirh** [comment]
Yes, drm isn't but I thought fbdev was fine again. Since 2022 activity has boomed, with like a thousand and then some commits.

Then, I could see the point for obsolescence if we were talking about ps1 (nommu and mips1) but the axe seems still pretty far from our target.

**Comment by frno7** [comment]
I agree that there’s opposition to the removal of [fbdev](https://en.wikipedia.org/wiki/Linux_framebuffer) etc. I oppose that myself. But I’m also observing an on-again, off-again, situation, where, in the somewhat near future, removal of critical components (all 32-bit architectures, for instance) seem inevitable.

Modern Linux isn’t the lean and efficient hobbyist kernel it once was in the nineties and early noughties. There are traces left, sure. These days there’s overwhelming pressure for enterprise features that have little or no practical use for hobbyists, and a corresponding purge of hobbyist features that are deemed deprecated. There’s considerable technical debt buried with some of that old stuff, sure, but to be fair, it’s likewise often old enterprise stuff.

Fighting an endless stream of kernel deprecation notices is frustrating, so why not look elsewhere? The choice I’ve made is to stay on Linux 5.4, as the best match of modernity, current compiler compatibility, and PlayStation 2 hardware. Of course, compiler compatibility may not last indefinitely, and there are other possibilities, such as switching to an entirely different kind of kernel, for example some BSD variant. Maybe that’s mostly a matter of taste, and maybe IOP device drivers could be made useful for both Linux and BSD kernels, to share maintenance, etc.

Still, I think Linux for PlayStation 2 is worthwhile, provided the problems mentioned with modern kernels are avoided. :-)
